### PR TITLE
Max: Tweaks on Extractions for the exporters

### PR DESCRIPTION
**Pull request by moonyuet**
## Changelog Description
With this PR
1. Suspend Refresh would be introduced in abc & obj extractors for optimization.
2. Allow users to choose the custom attributes to be included in abc exports

## Additional info
Not included the usd exporters

## Testing notes:
1. Create Model/PointCache/Camera Instance
3. Publish

### CODE REVIEW

File: openpype/hosts/max/api/lib.py [hunk 1]
```diff
@@ -509,3 +509,18 @@ def get_plugins() -> list:
...
+    """
+    rt.disableSceneRedraw()
+    rt.suspendEditing()
+
+    try:
+        yield
+
+    finally:
+        rt.enableSceneRedraw()
+        rt.resumeEditing()
```
Comment (BigRoy) [inline openpype/hosts/max/api/lib.py:526]: An issue in Maya implementation has been that similar logic failed in Maya batch / standalone. Should we check the same for Max whether this works fine or not in max standalone/batch mode?

If it doesn't work we might need to do some:

```python
if IS_HEADLESS:
    yield
    return
```

At the start of the context manager. I believe @antirotor did something recently in a maya PR.

File: openpype/hosts/max/plugins/create/create_pointcache.py [hunk 2]
```diff
@@ -9,3 +10,20 @@ class CreatePointCache(plugin.MaxCreator):
...
+            instance_data,
+            pre_create_data)
+
+    def get_pre_create_attr_defs(self):
+        attrs = super().get_pre_create_attr_defs()
+        return attrs + [
+            BoolDef("custom_attrs",
+                    label="Custom Attributes",
+                    default=False),
+        ]
```
Comment (BigRoy) [inline openpype/hosts/max/plugins/create/create_pointcache.py:29]: I wonder if it'd make more sense to expose this as an Attribute Definition on the Extractors instead? Just so it's much closer to what it's related to. Thoughts?

e.g.

```python
class ExtractAlembic(pyblish.api.InstancePlugin,
                     publish.OpenPypePyblishPluginMixin):

    def process(self, instance):
        attr_values = self.get_attr_values_from_data(instance.data)
        custom_attrs = attr_values.get("custom_attrs", False)
        # do something with the value

    def get_attribute_defs(self):
        return [
            BoolDef("custom_attrs",
                    label="Custom Attributes",
                    default=False),
        ]
```

**Review by LiborBatek — COMMENTED** [review]
Not tested if Custom Attribs got written but all ```Publish``` actions were successfuly done. And no any errors shown during publishing of those Instances from testing steps.

File: openpype/hosts/max/api/lib.py [hunk 3]
```diff
@@ -509,3 +509,20 @@ def get_plugins() -> list:
...
+
+@contextlib.contextmanager
+def suspended_refresh():
+    """Suspended refresh for scene and modify panel redraw.
+    """
+    rt.disableSceneRedraw()
+    rt.suspendEditing()
+    if is_headless():
+        yield
+        return
```
Comment (BigRoy) [inline openpype/hosts/max/api/lib.py:522]:
```suggestion
    if is_headless():
        yield
        return
    rt.disableSceneRedraw()
    rt.suspendEditing()
```

File: openpype/hosts/max/plugins/publish/extract_pointcache.py [hunk 4]
```diff
@@ -89,3 +85,54 @@ def process(self, instance):
...
+
+    def _set_abc_attributes(self, instance):
+        attr_values = self.get_attr_values_from_data(instance.data)
+        custom_attrs = attr_values.get("custom_attrs", False)
+        rt.AlembicExport.ArchiveType = rt.name("ogawa")
+        rt.AlembicExport.CoordinateSystem = rt.name("maya")
+        rt.AlembicExport.CustomAttributes = custom_attrs
+        rt.AlembicExport.UVs = True
+        rt.AlembicExport.VertexColors = True
+        rt.AlembicExport.PreserveInstances = True
```
Comment (BigRoy) [inline openpype/hosts/max/plugins/publish/extract_pointcache.py:138]: Is there a reason these need to be extracted earlier?

We should only need to override what we actually want to differ from the `ExtractAlembic` parent class.

```suggestion
class ExtractCameraAlembic(ExtractAlembic):
    """Extract Camera with AlembicExport."""
    
    label = "Extract Alembic Camera"
    families = ["camera"]


class ExtractModel(ExtractAlembic):
    """Extract Geometry in Alembic Format"""

    label = "Extract Geometry (Alembic)"
    families = ["model"]

    def _set_abc_attributes(self, instance):
        attr_values = self.get_attr_values_from_data(instance.data)
        custom_attrs = attr_values.get("custom_attrs", False)
        rt.AlembicExport.ArchiveType = rt.name("ogawa")
        rt.AlembicExport.CoordinateSystem = rt.name("maya")
        rt.AlembicExport.CustomAttributes = custom_attrs
        rt.AlembicExport.UVs = True
        rt.AlembicExport.VertexColors = True
        rt.AlembicExport.PreserveInstances = True
```

**Review by BigRoy — APPROVED** [review]
Didn't test it - but code changes look good to me. Very nice cleanup!

Only thing I wonder is whether we should rename the `extract_pointcache.py` file to `extract_alembic.py` since it now contains multiple extractors.

**Comment by mkolar** [comment]
> Didn't test it - but code changes look good to me. Very nice cleanup!
> 
> Only thing I wonder is whether we should rename the `extract_pointcache.py` file to `extract_alembic.py` since it now contains multiple extractors.

I agree completely. Good work, but it's not extracting just pointcache anymore.

**Review by LiborBatek — APPROVED** [review]
Publish went ok... without any issues. Tested in both legacy OP and Ayon mode.

image

**Pull request merged**